### DISCUSSION TITLE
Support Yarn Plug'n'Play

### DISCUSSION THREAD

**Discussion by Kurt-von-Laven** [avajs/ava#3132] [Q&A] [answered]
I am stumped trying to run AVA on [Yarn PnP](https://yarnpkg.com/features/pnp) and wondering how feasible it currently is, whether there may be some insight I missed, or how I might go about adding native support for PnP to AVA. I ran into chalk/chalk#531 and worked around it as described on that thread. That got AVA to load, but it [crashes on the first import](https://github.com/yarnpkg/berry/issues/3843#issuecomment-1146568693).

In PnP, most Node.js commands work by running `yarn <command>`, which [pretty much just runs `node --require "$(pwd)/.pnp.cjs" <command>`](https://yarnpkg.com/features/pnp#initializing-pnp). See the excerpt from Yarn's documentation below for a brief discussion of the purpose of `.pnp.cjs`. Hence, I tried passing `--require=./.pnp.cjs` via AVA's `nodeArguments` [config option](https://github.com/avajs/ava/blob/main/docs/06-configuration.md#options), but encountered the same crash as above on the first import.

> The `.pnp.cjs` file contains various maps: one linking package names and versions to their location on the disk and another one linking package names and versions to their list of dependencies. With these lookup tables, Yarn can instantly tell Node where to find any package it needs to access, as long as they are part of the dependency tree, and as long as this file is loaded within your environment (more on that in the next section). ~ [Fixing node_modules](https://yarnpkg.com/features/pnp#fixing-node_modules)

**Comment by novemberborn** [comment]
AVA follows the module system as defined by Node.js — it's up to Yarn to not break that.

**Comment by Kurt-von-Laven** [comment]
A fix for yarnpkg/berry#3843 was just merged, so anyone following this issue may want to try running AVA on the latest Yarn PnP once it's released.

**Comment by mrienstra** [comment]
> A fix for [yarnpkg/berry#3843](https://github.com/yarnpkg/berry/issues/3843) was just merged, so anyone following this issue may want to try running AVA on the latest Yarn PnP once it's released.

The fix is available in `4.0.0-rc.22`. Run [`yarn set version canary`](https://yarnpkg.com/cli/set/version) if you want to try it. v4 release is still a little ways off ("may take a couple more months"), but:
>[...] what's in master is stable, and I'd recommend you to try it. The only notable difference with stable is that we reserve the right to land a couple more breaking changes in future RCs, but in terms of stability it's almost always better to use RCs than stable.

... according to https://github.com/yarnpkg/berry/discussions/4895

See https://github.com/yarnpkg/berry/issues/3591 for v4 breaking changes. I've seen two mentions of people stumbling over `enableGlobalCache` default changing from `false` to `true`. ([Google search: site:yarnpkg.com "enableGlobalCache"](https://www.google.com/search?q=site%3Ayarnpkg.com+%22enableGlobalCache%22))

**Comment by mrienstra** [comment, marked as answer]
Update: the fix that adds support for the `package.json` `includes` field is also in the [Yarn `3.2.4` release](https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fcli%2F3.2.4).
